sampleSpecs/traits.smithy

```diff
@@ -0,0 +1,47 @@
+$version: "2.0"
+
+namespace smithy4s.example.traits
+
+/// A trait with no recursive references to itself
+@trait
+structure nonRecursiveTrait {}
+
+/// A trait with a direct recursive reference
+@trait
+@directRecursiveTrait
+structure directRecursiveTrait {}
+
+/// A trait with an indirect recursive reference
+@trait
+@indirectRecursiveTrait1
+structure indirectRecursiveTrait0 {}
+
+// trait that completes the recursion of the above, also recursive because it has to loop back through the above
+@trait
+@indirectRecursiveTrait0
+structure indirectRecursiveTrait1 {}
+
+/// A trait with no recursion in itself, referencing recursive traits
+@trait
+@nonRecursiveTrait
+@directRecursiveTrait
+@indirectRecursiveTrait0
+@indirectRecursiveTrait1
+structure nonRecursiveTraitReferencingOthers {}
+
+// TODO: this can't be generated because of https://github.com/disneystreaming/smithy4s/issues/1296
+// A trait that has direct recursive references to it via members.
+// @trait
+// structure directRecursiveViaMembersTrait {
+// @directRecursiveViaMembersTrait
+// member: String
+// }
+// 
```

Comment:
This is currently impossible to generate but I think the current state should render something like this:

```scala
implicit val schema: Schema[DirectRecursiveViaMembersTrait] = recursive(
  struct(
    string
      .optional[DirectRecursiveViaMembersTrait]("member", _.member)
      .addHints(smithy4s.example.traits.DirectRecursiveViaMembersTrait())
  ) {
    DirectRecursiveViaMembersTrait.apply
  }.withId(id).addHints(hints)
)
```

which is probably OK because of `recursive` laziness.

Reply:
aaand it's possible now. The generated result looks like this:

```scala
implicit val schema: Schema[DirectRecursiveViaMembersTrait] = recursive(
  struct(
    string
      .optional[DirectRecursiveViaMembersTrait]("member", _.member)
      .addHints(
        smithy4s.example.traits.DirectRecursiveViaMembersTrait(member = None)
      )
  ) {
    DirectRecursiveViaMembersTrait.apply
  }.withId(id).addHints(hints)
)
```

which is pretty much as expected.